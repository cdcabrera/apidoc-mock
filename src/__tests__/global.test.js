const global = require('../global');

describe('Global', () => {
  it('should return specific properties', () => {
    expect(Object.keys(global)).toMatchSnapshot('specific properties');
  });

  it('should minimally generate a consistent hash', () => {
    expect({
      valueObject: global.generateHash({ lorem: 'ipsum', dolor: ['sit', null, undefined, 1, () => 'hello world'] }),
      valueObjectConfirm:
        global.generateHash({ lorem: 'ipsum', dolor: ['sit', null, undefined, 1, () => 'hello world'] }) ===
        global.generateHash({ lorem: 'ipsum', dolor: ['sit', null, undefined, 1, () => 'hello world'] }),
      valueObjectConfirmSort:
        global.generateHash({ lorem: 'ipsum', dolor: ['sit', null, undefined, 1, () => 'hello world'] }) ===
        global.generateHash({ dolor: ['sit', null, undefined, 1, () => 'hello world'] }, { lorem: 'ipsum' }),
      valueInt: global.generateHash(200),
      valueFloat: global.generateHash(20.000006),
      valueNull: global.generateHash(null),
      valueUndefined: global.generateHash(undefined),
      valueArray: global.generateHash([1, 2, 3]),
      valueArraySort: global.generateHash([3, 2, 1]),
      valueArrayConfirmSort: global.generateHash([1, 2, 3]) !== global.generateHash([3, 2, 1]),
      valueSet: global.generateHash(new Set([1, 2, 3])),
      valueSetConfirmSort: global.generateHash(new Set([1, 2, 3])) === global.generateHash(new Set([3, 2, 1])),
      valueSymbol: global.generateHash(Symbol('lorem ipsum')),
      valueSymbolUndefined: global.generateHash(Symbol('lorem ipsum')) === global.generateHash(undefined),
      valueBoolTrue: global.generateHash(true),
      valueBoolFalse: global.generateHash(false)
    }).toMatchSnapshot('hash, object and primitive values');
  });

  it('should memoize function return values', () => {
    const testArr = [];
    let testStr;
    const testMemoReturnValue = global.memo(
      str => {
        const arr = ['lorem', 'ipsum', 'dolor', 'sit'];
        const randomStr = Math.floor(Math.random() * arr.length);
        const genStr = `${arr[randomStr]}-${str}`;
        testStr = genStr;
        testArr.push(genStr);
        return genStr;
      },
      { cacheLimit: 4 }
    );

    testMemoReturnValue('one');
    testMemoReturnValue('one');
    testMemoReturnValue('one');
    testMemoReturnValue('one');
    expect(testStr === testMemoReturnValue('one')).toBe(true);

    testMemoReturnValue('two');
    testMemoReturnValue('three');
    expect(testArr[0] === testMemoReturnValue('one')).toBe(true);
    expect(testArr[1] === testMemoReturnValue('two')).toBe(true);
    expect(testArr[2] === testMemoReturnValue('three')).toBe(true);
    expect(testArr[2] === testMemoReturnValue('three')).toBe(true);

    testMemoReturnValue('four');
    expect(testArr[3] === testMemoReturnValue('four')).toBe(true);
    expect(testArr.length).toBe(4);
  });

  it('should memoize async function return values', async () => {
    const asyncMemoValue = global.memo(value => new Promise(resolve => setTimeout(() => resolve(value), 10)));
    const asyncResponse = await asyncMemoValue('lorem ipsum');
    expect(asyncResponse).toBe('lorem ipsum');

    const asyncMemoError = global.memo(
      value => new Promise((_, reject) => setTimeout(() => reject(new Error(value)), 10))
    );
    await expect(async () => asyncMemoError('lorem ipsum')).rejects.toThrowErrorMatchingSnapshot('memoize async error');
  });

  it('should set a one-time mutable OPTIONS object', () => {
    const { OPTIONS } = global;

    // for testing set a consistent contextPath and parser
    OPTIONS.contextPath = '/';
    OPTIONS.apiDocBaseConfig.parsers.apimock = '/customParser.js';

    OPTIONS.lorem = 'et all';
    OPTIONS.dolor = 'magna';
    OPTIONS._set = {
      lorem: 'ipsum',
      sit: function () {
        return `function test ${this.contextPath}`;
      }
    };
    OPTIONS.lorem = 'hello world';
    OPTIONS.dolor = 'sit';

    expect({ isFrozen: Object.isFrozen(OPTIONS), OPTIONS }).toMatchSnapshot('immutable');
  });
});
