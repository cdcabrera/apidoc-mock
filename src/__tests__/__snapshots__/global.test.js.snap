// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global should memoize async function return values: memoize async error 1`] = `"lorem ipsum"`;

exports[`Global should minimally generate a consistent hash: hash, object and primitive values 1`] = `
{
  "valueArray": "0e8e2b3096e6aee20e3cd10b7788ce3d72502061",
  "valueArrayConfirmSort": true,
  "valueArraySort": "3c7765817d3c04f1104093994b6db4371672d3f7",
  "valueBoolFalse": "c0338288abd30f72cd65dad83833448f345f78fa",
  "valueBoolTrue": "d991ef01679b350a728c665383abb257cfeca53a",
  "valueFloat": "ba0abc40d5bc3a257ecf9396e3b2525d7fff76c9",
  "valueInt": "17dada6f0f5fe358d94f7b0849d17fc8b4c5ce1f",
  "valueNull": "f7fd41b8a32014e6ccd665899be623c6d5c99d43",
  "valueObject": "5a7036ff63f04efcc11d6b68eb97a4b9989d1a45",
  "valueObjectConfirm": true,
  "valueObjectConfirmSort": false,
  "valueSet": "86838cc5365d6e9d59fb005d2353dde173712a8a",
  "valueSetConfirmSort": true,
  "valueSymbol": "bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f",
  "valueSymbolUndefined": true,
  "valueUndefined": "bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f",
}
`;

exports[`Global should return specific properties: specific properties 1`] = `
[
  "contextPath",
  "generateHash",
  "memo",
  "OPTIONS",
]
`;

exports[`Global should set a one-time mutable OPTIONS object: immutable 1`] = `
{
  "OPTIONS": {
    "apiDocBaseConfig": {
      "dest": undefined,
      "dryRun": true,
      "parsers": {
        "apimock": "/customParser.js",
      },
      "silent": true,
      "src": undefined,
    },
    "contextPath": "/",
    "dolor": "magna",
    "lorem": "ipsum",
    "sit": "function test /",
  },
  "isFrozen": true,
}
`;
